### PR TITLE
Log viewer/scroll performance

### LINKED ISSUE #3991: Optimize the query for log messages to improve performance
- [ ] run a query to count the number of records to expect, and expand the time frame until we get to 1k logs
- [ ] add a max to the number of queries to run

### PR DESCRIPTION
**Pull request by bthesorceror**
Closes #3991

  - [x] Rebased/mergeable
  - [x] Tests pass
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: ui/src/logs/utils/index.ts [hunk 1]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
-  searchTerm: string | null = null
-) {
-  const {tags, areTagsAccepted} = config
+  searchTerm: string | null = null,
+  proxyLink: string,
+  namespace: Namespace
+): Promise<string> {
+  const parsedUpper = moment(upper)
+
+  let secondsBack = 30
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: this var name is unclear to me

File: ui/src/logs/utils/index.ts [hunk 2]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
+  let currentLower = parsedUpper.subtract(secondsBack, 'seconds')
+
+  while (true) {
+    if (secondsBack > SECONDS_AWAY_LIMIT) {
+      // One day
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: https://www.youtube.com/watch?v=WRmBChQjZPs

File: ui/src/logs/utils/index.ts [hunk 3]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
+  searchTerm: string | null = null,
+  proxyLink: string,
+  namespace: Namespace
+): Promise<string> {
+  const parsedUpper = moment(upper)
+
+  let secondsBack = 30
+  let currentLower = parsedUpper.subtract(secondsBack, 'seconds')
+
+  while (true) {
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: from my understanding, it seems like you're kind of constructing a generator here -- might an explicit generator be useful?

File: ui/src/logs/utils/index.ts [hunk 4]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
+      proxyLink,
+      namespace
+    )
+
+    if (count >= 400) {
+      break
+    }
 
-  return buildGeneralLogQuery(condition, config, filters, searchTerm)
+    secondsBack *= secondsBack
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: would you consider adding a code comment to provide context as to why this choice is being made?

File: ui/src/logs/utils/index.ts [hunk 5]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
+  namespace: Namespace
+): Promise<string> {
+  const parsedLower = moment(lower)
+
+  let secondsForward = 30
+  let currentUpper = parsedLower.add(secondsForward, 'seconds')
+
+  while (true) {
+    if (secondsForward > SECONDS_AWAY_LIMIT) {
+      break
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: any explicit UX we want to provide in this case?

File: ui/src/logs/utils/table.ts [hunk 6]
```diff
@@ -37,6 +38,9 @@ export const formatColumnValue = (
   switch (column) {
     case 'timestamp':
       return moment(+value / 1000000).format('YYYY/MM/DD HH:mm:ss')
+    case 'appname':
```
Comment (jaredscheib) [inline ui/src/logs/utils/table.ts]: is this related to the issue at hand?

File: ui/src/logs/utils/index.ts [hunk 7]
```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
...
+  lower,
+  upper,
+  config,
+  filters,
+  searchTerm,
+  proxyLink,
+  namespace
+): Promise<number> {
+  const {database, retentionPolicy, measurement} = config
+  const rpSegment = retentionPolicy ? `"${retentionPolicy}"` : ''
```
Comment (jaredscheib) [inline ui/src/logs/utils/index.ts]: ooh a ternany from @bthesorceror! ;)

File: ui/src/logs/actions/index.ts [hunk 8]
```diff
@@ -613,7 +656,7 @@ export const fetchMoreAsync = (queryTimeEnd: string) => async (
   }
 }
 
-export const fetchNewerAsync = (queryTimeStart: string) => async (
+export const fetchNewerLogsAsync = (queryTimeStart: string) => async (
```
Comment (jaredscheib) [inline ui/src/logs/actions/index.ts]: could you type the output?

File: ui/src/logs/actions/index.ts [hunk 9]
```diff
@@ -367,7 +368,23 @@ export const executeTableForwardQueryAsync = () => async (
...
+    const upper = await findForwardUpper(
+      time,
+      queryConfig,
+      filters,
+      searchTerm,
+      proxyLink,
+      namespace
+    )
+
+    const query: string = await buildInfiniteLogQuery(
```
Comment (jaredscheib) [inline ui/src/logs/actions/index.ts]: maybe `buildInfiniteScrollLogQuery`? otherwise it sounds like infinite could map to anything, such as the query, or that i can expect to return an infinite number of logs

File: ui/src/logs/utils/index.ts (hunk 3 above)
Comment (bthesorceror) [reply ui/src/logs/utils/index.ts]: Don't think it would quite fit and as the forward and backward differ slightly I don't think we get a lot of benefit out of extraction quite yet.

File: ui/src/logs/actions/index.ts [hunk 10]
```diff
@@ -595,7 +628,17 @@ export const fetchMoreAsync = (queryTimeEnd: string) => async (
   const params = [namespace, proxyLink, tableQueryConfig]
 
   if (_.every(params)) {
-    const query = buildBackwardLogQuery(
+    const queryTimeStart = await findOlderLowerTimeBounds(
```
Comment (jaredscheib) [inline ui/src/logs/actions/index.ts]: does this merit a type?

File: ui/src/logs/actions/index.ts [hunk 11]
```diff
@@ -595,7 +628,17 @@ export const fetchMoreAsync = (queryTimeEnd: string) => async (
...
+      queryTimeEnd,
+      newQueryConfig,
+      filters,
+      searchTerm,
+      proxyLink,
+      namespace
+    )
+
+    const query = await buildInfiniteScrollLogQuery(
```
Comment (jaredscheib) [inline ui/src/logs/actions/index.ts]: does this merit a type?

File: ui/src/logs/actions/index.ts [hunk 12]
```diff
@@ -489,8 +522,8 @@ export const executeHistogramQueryAsync = () => async (
 
 export const executeTableQueryAsync = () => async (dispatch): Promise<void> => {
   await Promise.all([
-    dispatch(executeTableForwardQueryAsync()),
-    dispatch(executeTableBackwardQueryAsync()),
+    dispatch(executeTableNewerQueryAsync()),
```
Comment (jaredscheib) [inline ui/src/logs/actions/index.ts]: would Results be appropriate within this func name for consistency with `executeTableOlderResultsQueryAsync`?

**Comment by jaredscheib** [comment]
Looks great to me -- thanks @bthesorceror -- much clearer. Thanks! Left 3 last minor thoughts, and I meant to Approve in my latest review.

File: ui/src/logs/actions/index.ts (hunk 10 above)
Comment (bthesorceror) [reply ui/src/logs/actions/index.ts]: as the functions have a return type I don't think so

File: ui/src/logs/actions/index.ts (hunk 11 above)
Comment (bthesorceror) [reply ui/src/logs/actions/index.ts]: as the functions have a return type I don't think so
